components/eam/src/physics/crm/pam/pam_driver.cpp

```diff
@@ -14,26 +13,90 @@
 #include "surface_friction.h"
 #include "scream_cxx_interface_finalize.h"
 
-#include "pam_hyperdiffusion.h"
-
 // Needed for p3_init
 #include "p3_functions.hpp"
 #include "p3_f90.hpp"
 
 #include "pam_debug.h"
 bool constexpr enable_check_state = false;
 
-extern "C" void pam_driver() {
+
+inline int pam_driver_set_subcycle_timestep( pam::PamCoupler &coupler, real crm_dt_fixed ) {
+  // calculate the CFL condition and adjust the PAM time loop subcylcing
   //------------------------------------------------------------------------------------------------
-  using yakl::intrinsics::abs;
-  using yakl::intrinsics::maxval;
-  using yakl::atomicAdd;
   using yakl::c::parallel_for;
   using yakl::c::SimpleBounds;
+  using yakl::atomicMax;
+  //------------------------------------------------------------------------------------------------
+  auto nens          = coupler.get_option<int>("ncrms");
+  auto gcm_nlev      = coupler.get_option<int>("gcm_nlev");
+  auto crm_nz        = coupler.get_option<int>("crm_nz");
+  auto crm_nx        = coupler.get_option<int>("crm_nx");
+  auto crm_ny        = coupler.get_option<int>("crm_ny");
+  auto crm_dx        = coupler.get_option<real>("crm_dx");
+  auto crm_dy        = coupler.get_option<real>("crm_dy");
+  auto &dm_device = coupler.get_data_manager_device_readonly();
+  auto &dm_host   = coupler.get_data_manager_host_readonly();
+  auto uvel       = dm_device.get<real const,4>("uvel");
+  auto wvel       = dm_device.get<real const,4>("wvel");
+  auto input_zint = dm_host.get<real const,2>("input_zint").createDeviceCopy();
+  //------------------------------------------------------------------------------------------------
+  yakl::ParallelMax<real,yakl::memDevice> pmax( crm_nz*nens );
+  real cfl = 0;
+  int num_subcycle = 1;
+  int constexpr max_num_subcycle = 10;
+  real2d wvel_max("wvel_max",crm_nz,nens);
+  real2d uvel_max("uvel_max",crm_nz,nens);
+  real2d cfl_max("cfl_max",  crm_nz,nens);
+  //------------------------------------------------------------------------------------------------
+  // initialize max U and W arrays
+  parallel_for( SimpleBounds<2>(crm_nz,nens) , YAKL_LAMBDA (int k, int n) {
+    wvel_max(k,n) = 0.0;
+    uvel_max(k,n) = 0.0;
+  });
+  // calculate max U and W
+  parallel_for( SimpleBounds<4>(crm_nz,crm_ny,crm_nx,nens) , YAKL_LAMBDA (int k, int j, int i, int n) {
+    yakl::atomicMax(uvel_max(k,n), sqrt(uvel(k,j,i,n)*uvel(k,j,i,n)) );
+    yakl::atomicMax(wvel_max(k,n), fabs(wvel(k,j,i,n)) );
+  });
+  // find max CFL between horizontal and vertical CFL values
+  parallel_for( SimpleBounds<2>(crm_nz,nens) , YAKL_LAMBDA (int k, int n) {
+    int k_gcm = gcm_nlev-1-k;
+    real crm_dz = input_zint(k_gcm,n) - input_zint(k_gcm+1,n);
+    real cfl_u = uvel_max(k,n)*crm_dt_fixed/crm_dx;
+    real cfl_w = wvel_max(k,n)*crm_dt_fixed/crm_dz;
+    cfl_max(k,n) = max(cfl_u,cfl_w);
+  });
+  // calculate final CFL across ensemble
+  real cfl_loc = pmax(cfl_max.data());
```

Comment:
You can now say the following here:
```C++
real cfl_loc = yakl::intrinsics::maxval(cfl_max);
```
Much clearer and easier :)

Reply:
Excellent, just made the fix.